Rename SelfAux to ProjectionAux

GiudGiud opened the pull request:
Adds tests for projecting as well

TODO (potential) :
- [x] add warning or info message for inexact projections due to shape/orders
- [ ] add warning or info message for inexact projections due to insufficient quadrature
- [x] add error on unsupported FE families for source variable (if any)
- [x] add error on unsupported FE families for target variable
- [x] investigate SZABAB higher order 1d & 2d projection issue
- [x] investigate higher order side_hierarchic 1d & 2d projection issue
- [x] investigate monomial -> lagrange 1d projection issue
- [ ] understand all that can be wrong with this approach
- [x] 2d tests on quads
- [x] 2d tests on tris
- [x] valgrind the memory error for 2D

my current take on the errors is:
- side_hierarchic isnt supported. Setting dofs indirectly through auxkernels simply doesnt work for these variables
- szabab, lagrange and other nodal variables needed a projection from elemental to nodal
- nodal higher order was problematic because the lower order nodal source didnt have the right value at the center node (indexing at _qp=0 but _qp = 0 for an edge node)

Higher order Szabab still looks odd. Where are the dofs for those?
<img width="911" alt="Screen Shot 2023-05-01 at 2 56 11 PM" src="https://user-images.githubusercontent.com/25037151/235529756-42b6a1ec-82ca-413e-98b7-d63e3123ea03.png">

GiudGiud committed:
Rename SelfAux to ProjectionAux
refs #22373

GiudGiud committed:
Add 1d test for projections between FE families

GiudGiud committed:
Upgrade documentation
Make coupled variable mandatory

GiudGiud commented on test/tests/auxkernels/projection_aux/1d.i:
```diff
@@ -0,0 +1,497 @@
+  # LAGRANGE, MONOMIAL, HERMITE, SCALAR, HIERARCHIC, CLOUGH, XYZ, SZABAB, BERNSTEIN,
+  # L2_LAGRANGE, L2_HIERARCHIC, RATIONAL_BERNSTEIN, SIDE_HIERARCHIC
+  # Notes:
+  # - 'elemental': MONOMIAL, XYZ, L2_LAGRANGE, L2_HIERARCHIC
+  # - 'nodal': LAGRANGE, HERMITE, HIERARCHIC, CLOUGH, SZABAB, BERNSTEIN, RATIONAL_BERNSTEIN
+  # - Hermite, Hierachic, Clough, Bernstein, rational Berstein cannot be created in 1D EDGE3
+  # - Clough, Szabab, L2_lagrange, L2_hierarchic cannot be created as constant
```
@roystgnr is any of this not true?

GiudGiud replied:
expanded the list in current version of file

GiudGiud committed:
Add test for quad9 projections, refs #22373

GiudGiud committed:
Add test for TRI6 projections, refs #22373

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1495352/) on 3f51cb6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24237/site)

This comment will be updated on new commits.

GiudGiud commented:
so what we need here is a projection rule from elemental to nodal variables, as the current logic isnt valid, and fails silently once the right boolean is set to allow elemental variables coupled into nodal.

GiudGiud committed:
Add projection capability from elemental to nodal variables, #22373

GiudGiud committed:
Add messages to user in ProjectionAux:
- unexpected projection needed for nodal order
- unsupported side hierarchic

GiudGiud committed:
Remove unsupported side hierarchic

GiudGiud committed:
Add test failures from making SelfAux into more potent ProjectionAux

GiudGiud committed:
Add node ghosting functor for parallel projection, refs #22373

GiudGiud commented on framework/src/auxkernels/ProjectionAux.C:
```diff
@@ -0,0 +1,95 @@
+      {
+        rm_params.set<unsigned short>("layers") = obj_params.get<unsigned short>("ghost_layers");
+        rm_params.set<bool>("use_displaced_mesh") = obj_params.get<bool>("use_displaced_mesh");
+      });
```
it s absolutely crazy that it s all it took to get mpi working

GiudGiud commented on framework/src/auxkernels/SelfAux.C:
```diff
@@ -1,35 +0,0 @@
-//* This file is part of the MOOSE framework
```
I git mved, but git still thinks i deleted

lindsayad commented on framework/src/auxkernels/ProjectionAux.C:
```diff
@@ -0,0 +1,95 @@
+    return _v[_qp];
+  // Handle elemental variable projection into a nodal variable
+  // AND nodal low order -> nodal higher order
+  else
```
The `else` cases made me think of nodal patch recovery. @hugary1995 am I wrong for thinking of that?

hugary1995 replied:
Ya, both are least squares fitting essentially. My understanding is that auxkernel doesn't construct a larger patch of element -- it's just extrapolating into the nodes, whereas NPR does the fitting over all nodal neighbors.

GiudGiud committed:
Add better projection for continuous elemental variables

GiudGiud committed:
Remove references to deprecated SelfAux

lindsayad commented:
I'm going to defer to @roystgnr on this one and unsubscribe. If you need me, just mention me

GiudGiud commented on framework/src/auxkernels/ProjectionAux.C:
```diff
@@ -0,0 +1,103 @@
+    {
+      const auto & elem = _mesh.elemPtr(id);
+      elem_values[index] =
+          _source_sys.system().point_value(_source_variable->number(), elem->true_centroid(), elem);
```
going to have to add something here when the variable isnt defined on that element

GiudGiud replied:
done

GiudGiud committed:
Add treatment for block restriction: just skip elements

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1498684/) on 6a935ae wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7e596c</th>
      <th colspan=4>#24237 6a935a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.16%</td>
      <td>85.16%</td>
      <td>-0.00%</td>
      <td>82.69%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>87139</td>
      <td>87172</td>
      <td>+33</td>
      <td>43</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15183</td>
      <td>15192</td>
      <td>+9</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24237/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24237/coverage/framework)


## Modules coverage

### Chemical reactions

<table>
  <thead>
    <tr>
      <th></th>
      <th>7e596c</th>
      <th colspan=4>#24237 6a935a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>92.33%</td>
      <td>92.35%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1131</td>
      <td>1134</td>
      <td>+3</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>94</td>
      <td>94</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24237/coverage_diff/chemical_reactions)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24237/coverage/chemical_reactions)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24237/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 82.69% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
Passing now

GiudGiud commented:
so at this point I believe this code in AuxKernel is not fit for **setting** variables that have DOFs on elements and nodes
```
  if (isNodal()) /* nodal variables */
  {
    if (_var.isNodalDefined())
    {
      _qp = 0;
      ComputeValueType value = computeValue();
      // update variable data, which is referenced by other kernels, so the value is up-to-date
      _var.setNodalValue(value);
    }
  }
```

isNodal() ends up relying on:
```
  _is_nodal = (_continuity == C_ZERO || _continuity == C_ONE);
```
in MooseVariableData.C
which imo is pretty ad-hoc.

I think isNodal could become either, to be clearer,
- isOnlyNodal
- hasOnlyDoFsOnNodes

In the meantime, I would like for this PR to go under review as it's needed to support an ANL project

thoughts?

roystgnr reviewed (changes requested):
If you're in a hurry we can just add a mooseError for all the non-supported target variable types ... but I really want to at least figure out what's going wrong in the p=0 `MONOMIAL` to p=1 `LAGRANGE` case before we merge this.

roystgnr commented on test/tests/auxkernels/projection_aux/gold/1d_linear.csv:
```diff
@@ -0,0 +1,2 @@
+time,base_elem_proj_fv,base_elem_proj_l2_hierarchic,base_elem_proj_l2_hierarchic_high,base_elem_proj_l2_lagrange,base_elem_proj_l2_lagrange_high,base_elem_proj_lagrange,base_elem_proj_lagrange_high,base_elem_proj_mono,base_elem_proj_mono_high,base_elem_proj_szabab,base_elem_proj_szabab_high,base_elem_proj_xyz,base_elem_proj_xyz_high,base_nodal_proj_fv,base_nodal_proj_l2_hierarchic,base_nodal_proj ...
+1,0.057735026918963,4.5865336372792e-16,1.1210259799054e-15,4.5865336372792e-16,6.0402662910232e-16,0.025819888974716,0.016329931618554,0.057735026918963,7.3420361746659e-16,0.025819888974716,1.3689467250025,0.057735026918963,1.0169976882749e-15,0.086602540378444,7.3494938060193e-16,1.6897450165163e-15,7.3494938060193e-16,7.3420361746659e-16,0,0.024494897427832,0.086602540378444,8.5486371865671e- ...
```
Never mind; I think I get it now.  We're evaluating `ElementL2Difference` here, not from the exact IC, but from the `base_elem` approximation to the exact IC.  Even with an exact Lagrange solution we'd still be seeing the error in `base_elem`.

roystgnr replied:
No, no, that's still not the case.  You're bumping up `AuxVariables/base_elem/order=FIRST` in `tests`/`1d_linear`.  That should still recover the exact solution.

GiudGiud replied:
>  but we have 0.0258 error when projecting to a linear LAGRANGE?

that was from using the centroid value near a boundary. It was basically forcing a flat profile near the boundary

> we can chalk up p=2 SZABAB to our previous WTF ...

all nodal variables unsupported by the auxkernel system are disallowed now

GiudGiud committed:
Add back SelfAux as CopyValueAux. It s still clearer than a Projection if the variable is the same
Found another one in the THM module... this needed to happen

GiudGiud commented on framework/src/auxkernels/ProjectionAux.C:
```diff
@@ -0,0 +1,113 @@
+  params.addParam<unsigned short>("ghost_layers", 1, "The number of layers of elements to ghost.");
+  params.addRelationshipManager(
+      "ElementPointNeighborLayers",
+      Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC,
```
changed to algebraic. I thought it made sense to have geometric when you need element info and algebraic as well for variable info

roystgnr replied:
It does make sense!  So much sense that we don't ever do the latter without the former.

lindsayad replied:
Maybe I shouldn't have unsubscribed. We don't add MOOSE algebraic ghosting functors to the mesh, so the `|` is needed

lindsayad replied:
So this works because of the default geometric ghosting in libMesh (and also because of Exodus output turned on in this input file). If I disable the default geometric ghosting in libMesh (which we never do in MOOSE so this is somewhat of a hypothetical exercise) and turn off exodus output, then I get some exodiffs with distributed mesh

lindsayad replied:
I don't understand why the heck I added `MooseGhostPointNeighborLayers` when we have `ElementPointNeighborLayers`. Opened https://github.com/idaholab/moose/issues/24573 to remove it

roystgnr replied:
>We don't add MOOSE algebraic ghosting functors to the mesh

And ... of course we don't.  `/*to_mesh = */ false`.  I'd forgotten that I'd given up on fighting the `break_encapsulation_because_thats_easier_to_optimize` option; sorry for both stages of that.

GiudGiud commented on framework/src/auxkernels/ProjectionAux.C:
```diff
@@ -0,0 +1,113 @@
+
+    // Get the neighbor element centroid values & element volumes
+    std::vector<Real> elem_values(elem_ids->second.size(), 0);
+    std::vector<Real> elem_volumes(elem_ids->second.size(), 0);
```
yup

GiudGiud committed:
Address Roy's review
- use reference for variable class member
- use algebraic only ghosting
- accumulate variable values faster

GiudGiud committed:
Use nodal values for low to high order nodal projections

GiudGiud commented on framework/src/auxkernels/ProjectionAux.C:
```diff
@@ -0,0 +1,113 @@
+      if (_source_variable->hasBlocks(elem->subdomain_id()))
+      {
+        elem_values[index] = _source_sys.system().point_value(
+            _source_variable->number(), elem->true_centroid(), elem);
```
Umm you're right.
I added the second option last, and did not think to move that case there.

In my mind it was fair to weight the centroid value with the volume. But for using the (discontinuous) nodal value, I would have to weight with either something like the local solid angle of the element or some distance to the centroid instead of the volume
I changed it. I just want a projection rule, it will have to be approximate anyway

GiudGiud commented on test/tests/auxkernels/projection_aux/gold/1d_constant.csv:
```diff
@@ -0,0 +1,2 @@
+time,base_elem_proj_fv,base_elem_proj_l2_hierarchic,base_elem_proj_l2_hierarchic_high,base_elem_proj_l2_lagrange,base_elem_proj_l2_lagrange_high,base_elem_proj_lagrange,base_elem_proj_lagrange_high,base_elem_proj_mono,base_elem_proj_mono_high,base_elem_proj_szabab,base_elem_proj_szabab_high,base_elem_proj_xyz,base_elem_proj_xyz_high,base_nodal_proj_fv,base_nodal_proj_l2_hierarchic,base_nodal_proj ...
+1,0,6.9668399202232e-16,1.7458725287266e-15,6.9668399202232e-16,6.8558706426988e-16,2.0934566115784e-16,0,0,8.6315427321852e-16,2.0934566115784e-16,1.7888543819998,0,8.6315427321852e-16,1.4043333874307e-16,6.5535558080098e-16,1.4066719958574e-15,6.5535558080098e-16,7.4014868308344e-16,0,1.4043333874307e-16,1.4043333874307e-16,8.4778531236866e-16,0,1.5652475842499,1.4043333874307e-16,1.12638860489 ...
```
> Any idea why the p=2 SZABAB cases are failing here?

AuxKernels only set the nodal value. None of the additional dofs. 
Same for all the others

GiudGiud committed:
Use volume-weighted nodal values for projection of nodal-discontinuous variables

GiudGiud committed:
Forbid higher order nodal variables that have more than just 1 value dof per node from being set by auxkernels

GiudGiud committed:
Remove higher order nodal variables (that are not lagrange) from projection test

GiudGiud commented:
not ready yet, trying to change something in THM now

GiudGiud committed:
Use ProjectionAux in THM to go from lagrange flow areas to const monomial

GiudGiud commented:
ok done now! Turns out it was an easy fix with @joshuahansel 's insights

Ready for re-review
@joshuahansel

GiudGiud commented:
Of course Griffin uses both the SelfAux and their own CopyAux...

GiudGiud committed:
Use ProjectionAux for deprecation instead, downstream apps have been projecting like crazy without knowing

GiudGiud committed:
Add a test for CopyValueAux since it's not tested through deprecation of SelfAux

moosebuild commented:
Job [Framework 1](https://civet.inl.gov/job/1558829/) on f5e6c31 : invalidated by @GiudGiud

unrelated

GiudGiud commented:
Green now

roystgnr reviewed (approved):
I'm not seeing any more problems (or at least nothing that I'd call a bug rather than a hard-to-add missing feature...)

GiudGiud commented:
Awesome! Thanks a lot for the review! This has bugged me for a long time now.

lindsayad commented on test/tests/auxkernels/projection_aux/1d.i:
```diff
@@ -0,0 +1,442 @@
+  # L2_LAGRANGE, L2_HIERARCHIC, RATIONAL_BERNSTEIN, SIDE_HIERARCHIC
+  # Notes:
+  # - 'elemental': MONOMIAL, XYZ, L2_LAGRANGE, L2_HIERARCHIC
+  # - 'nodal': LAGRANGE, HERMITE, HIERARCHIC, CLOUGH, SZABAB, BERNSTEIN, RATIONAL_BERNSTEIN
```
Probably not

GiudGiud commented:
@yukkiezy and @aprilnovak
